crates/sui-indexer/src/handlers/tx_processor.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO remove the dead_code attribute after integration is done
+#![allow(dead_code)]
+
+use async_trait::async_trait;
+use move_binary_format::CompiledModule;
+use move_core_types::language_storage::ModuleId;
+use mysten_metrics::monitored_scope;
+use mysten_metrics::spawn_monitored_task;
+use tokio::sync::watch;
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use sui_types::object::Object;
+use tokio::time::Duration;
+use tokio::time::Instant;
+
+use sui_json_rpc::get_balance_changes_from_effect;
+use sui_json_rpc::get_object_changes;
+use sui_json_rpc::ObjectProvider;
+use sui_types::base_types::SequenceNumber;
+use sui_types::digests::TransactionDigest;
+use sui_types::effects::{TransactionEffects, TransactionEffectsAPI};
+use sui_types::transaction::{TransactionData, TransactionDataAPI};
+use tracing::debug;
+
+use sui_types::base_types::ObjectID;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+
+use crate::errors::IndexerError;
+use crate::metrics::IndexerMetrics;
+
+use crate::types_v2::IndexedPackage;
+use crate::types_v2::{IndexedObjectChange, IndexerResult};
+
+/// An in-mem cache for packages. It has static lifetime.
+/// Since we batch process checkpoints, it's possible that when a package
+/// is looked up, it has not been persisted in the database yet. So it
+/// works as an in-mem store for package resolution.
+/// To avoid bloating memory, we GC packages that are older than the committed checkpoints.
```

Comment:
Tracking based on CheckpointSeq number may not be the best way to handle GCing this for a couple of reasons:
- We likely want to use this module resolution package cache on the read path which wouldn't have an analogous GC trigger.
- If we ever have to fallback to reading a package from the DB we likely want to cache that read for a period of time since we'll likely want to use the same module again and again. Think about a RPC read request that wants to load N events from 1 transaction, where all those events are the same type. In the current implementation I think we'd need to fallback to read the pkg N times vs just once.

We could likely do something where we can just invalidate the whole cache every so often. Although on the read side one challenge for this would be the system package upgrading, but maybe if we GC often enough (idk every 30min or something + maybe around epoch changes) then we could be fine.

Reply:
...I would just stick this behind an LRUCache and call it a day, which is I think roughly what we do today on nodes.